### DISCUSSION TITLE
Concerning slack

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [google/or-tools#2298] [Routing (and legacy CP) questions]
Hopefully a fairly straightforward question concerning slack. Is that always a [firm equivalance](https://developers.google.com/optimization/routing/dimensions#slack_variables), that is `slack(i) = cumul(j) - cumul(i) - transit(i, j)`? Or rather, is that, _at most_, i.e. `slack(i) <= cumul(j) - cumul(i) - transit(i, j)`?

**Comment by lperron** [comment]
Slack is exact.

Le mer. 23 déc. 2020 à 18:06, Michael W Powell <notifications@github.com> a
écrit :

**Reply by mwpowellhtx** [reply]
@lperron I wonder, is _slack_ a misnomer in that case? Taken in the context of a [rope](https://dictionary.com/browse/slack). Else, how would one emulate that?

**Comment by lperron** [comment]
That is the usual name in or, especially the mip community.

Le mer. 23 déc. 2020 à 19:33, Michael W Powell <notifications@github.com> a
écrit :

**Reply by mwpowellhtx** [reply]
As terminology goes, fair enough. As actual behavior goes, a misnomer.

**Comment by Mizux** [comment]
https://github.com/google/or-tools/blob/b77bd3ac69b7f3bb02f55b7bab6cbb4bab3917f2/ortools/constraint_solver/routing.cc#L6151-L6165

https://github.com/google/or-tools/blob/b77bd3ac69b7f3bb02f55b7bab6cbb4bab3917f2/ortools/constraint_solver/constraint_solver.h#L1602

so basically you'll have:
```
cumul[next(i)] == cumul[i] + transits[i] == cumul[i] + fixed_transit[i] + slack[i] + dependent_transit[i]
with:
* fixed_transit[i] is the return of your `transit(from i, to next(i))` callback you register
* dependent_transit[i] usually null (only used with `AddDimensionDependentDimension()` which is not binded)
```

**Reply by mwpowellhtx** [reply]
@lperron @Mizux In keeping with industry terminology, yes, I grasp that much. However, as a lay expectation of definition of _slack_... Wouldn't it make better sense to consider that? Absent that when _adding dimension_, how would you emulate that? _Set some cumul variable ranges_?
